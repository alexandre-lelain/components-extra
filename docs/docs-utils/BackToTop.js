import styled from 'styled-components'

import RealBackToTop from 'components/BackToTop'

export const BackToTop = styled(RealBackToTop)`
  position: relative;
  margin: auto;
  display: flex;
  margin-top: 48px;
`

export const ExtendedBackToTop = styled(BackToTop)`
  background-color: green;
  padding: 38px;
  border-radius: 12px;
  box-shadow: 2px 2px 2px 2px grey;
`