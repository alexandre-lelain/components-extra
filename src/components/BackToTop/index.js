import React, { forwardRef, useEffect, useState } from 'react'
import styled from 'styled-components'

import ArrowUpward from './components/ArrowUpward'
import Button from './components/Button'

const START_HEIGHT = 20

const BackToTop = ({ className, forwardedRef = null }) => {
  const [display, setDisplay] = useState(false)
  const { body, documentElement } = document

  useEffect(() => {
    const onScroll = () => {
      if (body.scrollTop > START_HEIGHT || documentElement.scrollTop > START_HEIGHT) {
        setDisplay(true)
      } else {
        setDisplay(false)
      }
    }

    document.addEventListener('scroll', onScroll)
    return () => document.removeEventListener('scroll', onScroll)
  }, [body.scrollTop, documentElement.scrollTop])

  const scrollToTop = () => documentElement.scrollIntoView({ behavior: 'smooth' })

  return (
    <Button
      aria-label="Back to top"
      className={className}
      color="primary"
      isDisplayed={display}
      onClick={scrollToTop}
      ref={forwardedRef}
    >
      <ArrowUpward color="secondary" />
    </Button>
  )
}

BackToTop.displayName = 'BackToTop'

export { BackToTop, START_HEIGHT }
export default styled(forwardRef((props, ref) => <BackToTop {...props} forwardedRef={ref} />))``
