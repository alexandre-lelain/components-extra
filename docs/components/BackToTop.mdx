---
name: BackToTop
menu: Components
route: /components/back-to-top
---

import { Playground } from 'docz'
import { BackToTop, Extended, ExtendedBackToTop, Props } from '../docs-utils'

# BackToTop

<br />

This is the component you can see at the bottom-right corner of your window when you scroll down.

> Note: clicking on the button in the stories will bring you back to the top of the window ;)

## Default

<Playground>
  <BackToTop />
</Playground>

## Small

<Playground>
  <BackToTop size="small"/>
</Playground>

## Extended

<Extended name="BackToTop">
  <ExtendedBackToTop />
</Extended>

## Props

This component extends `Material's UI Fab` so you can use all its props. Check out its [Api](https://material-ui.com/api/fab/).

<Props of={BackToTop} />

> No props needed to use the component as is! But you can use the ones above :)