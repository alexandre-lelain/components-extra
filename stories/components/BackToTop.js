import React from 'react'
import Typography from '@material-ui/core/Typography'
import { storiesOf } from '@storybook/react'

import { BackToTop } from 'components/BackToTop'
import createApiStory from '../helpers/createApiStory'

const story = storiesOf('Components|BackToTop', module)

Typography.displayName = 'p'

story.add('Default', () => (
  <>
    <Typography>Scroll down... ;) </Typography>
    <div style={{ height: '1800px' }}></div>
    <BackToTop />
  </>
))

createApiStory(story, BackToTop)
