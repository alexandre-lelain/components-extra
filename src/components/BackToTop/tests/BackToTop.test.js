import React, { useRef, useLayoutEffect } from 'react'
import renderer from 'react-test-renderer'

import BackToTop from '..'

describe('BackToTop testing', () => {
  let testRef = { current: undefined }

  const TestBackToTop = () => {
    const ref = useRef()

    useLayoutEffect(() => {
      testRef = ref
    }, [])

    return <BackToTop ref={ref} />
  }

  const component = renderer.create(<TestBackToTop />)
  const tree = component.toJSON()

  test('BackToTop button renders correctly', () => {
    expect(tree).toMatchSnapshot()
  })

  test('BackToTop button forwards the given ref correctly.', () => {
    expect(testRef.current).not.toBeUndefined()
  })
})
