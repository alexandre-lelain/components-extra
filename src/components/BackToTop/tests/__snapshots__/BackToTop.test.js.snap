// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop testing BackToTop button renders correctly 1`] = `
<button
  aria-label="Back to top"
  className="MuiButtonBase-root MuiFab-root sc-bwzfXH uTxCW sc-bdVaJa cupWFY MuiFab-primary"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiFab-label"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
      />
    </svg>
  </span>
</button>
`;
